### ISSUE TITLE
Add `matchmissing = :notequal` option

### ISSUE THREAD

**Issue by nilshg** [JuliaData/DataFrames.jl#2650]
Consider two DataFrames:

```
julia> df = DataFrame(a = [1, 2, missing, missing])
4×1 DataFrame
 Row │ a       
     │ Int64?  
─────┼─────────
   1 │       1
   2 │       2
   3 │ missing 
   4 │ missing 

julia> df2 = DataFrame(a = [1, 3, missing, missing], b = rand(4))
4×2 DataFrame
 Row │ a        b        
     │ Int64?   Float64  
─────┼───────────────────
   1 │       1  0.459054
   2 │       3  0.649346
   3 │ missing  0.875563
   4 │ missing  0.709856
```

currently to join them I would need to set `matchmissing = :equal`, which produces duplicates:

```
julia> leftjoin(df, df2, on = :a, matchmissing = :equal)
6×2 DataFrame
 Row │ a        b
     │ Int64?   Float64?       
─────┼─────────────────────────
   1 │       1        0.459054
   2 │ missing        0.875563
   3 │ missing        0.875563
   4 │ missing        0.709856
   5 │ missing        0.709856
   6 │       2  missing     
```

I would like an option `matchmissing = :ignore` (or whatever other name) that preserves the left table exactly, and only adds information on the right side where non-missing values match. Currently I think this would be achieved via

```
julia> leftjoin(df, dropmissing(df2), on = :a, matchmissing = :equal)
4×2 DataFrame
 Row │ a        b
     │ Int64?   Float64?       
─────┼─────────────────────────
   1 │       1        0.459054
   2 │       2  missing        
   3 │ missing  missing        
   4 │ missing  missing       
```

which is a bit counterintuitive from an API perspective (I need to set `matchmissing` to `:equal` even though I don't want to match missings!), and might also suboptimal from an efficiency perspective.

**Comment by bkamins** [comment]
I think we should add it.

**Comment by pstorozenko** [comment]
I believe
`leftjoin(df, df2, on = :a, matchmissing = :ignore)`
should work like
`leftjoin(df, dropmissing(df2, :a), on = :a, matchmissing = :equal)`
instead of proposed
`leftjoin(df, dropmissing(df2), on = :a, matchmissing = :equal)`
right?
It seems not working properly for `df2` with missings in `b` column.

How about other types of join?
Rightjoin is symmetric to leftjoin.
With inner and outer I think we should drop missings on key in both dataframes.
With anti and semi, we should only drop missings in df2.

What do you think?

**Comment by bkamins** [comment]
`dropmissing(df2, :a)` is a correct way to write it. I think that @nilshg just made a typo.

Regarding other joins - what you propose makes sense. However, we need to make sure that these rules are easy to communicate as the following is also feasible:

`leftjoin` and `rightjoin` - make sense as they promise not to touch left or right table respectively.
`innerjoin` - makes sense - as then we say (consistently with left and right join above) that `missing` is never considered a match to `missing` but does not error.
`outerjoin` is tricky - the other approach to your would be to give the following logic: `outerjoin` promises to keep all rows from both tables; this means that we cannot do `dropmissing`; then row with `missing` in key would be "no-match" but still included (but this is not a super nice solution); the question is if this would be useful; we can always say that `:ignore` is not supported by `outerjoin` and decide later what we prefer.
`semijoin` - makes sense; `missing` is dropped as it does not match `df2`
`antijoin` - makes sense; `missing` is retained as it does not match `df2.

So in summary - the problematic case is `outerjoin`, but as commented above we can skip support of `:ignore` in this case to avoid user confusion.

And the rules would be:
* `:equal` => assume that `missing == missing` is `true`;
* `:ignore` => assume that `missing == missing` is `false`;

**Comment by pstorozenko** [comment]
This is my very first contribution to any open source project (as well as developing pkg for julia) so I don't know how to share my findings but I'll give it a try.
I've implemented `:ignore` by passing `kind` variable to `DataFrameJoiner` and adding the following to the `DataFrameJoiner` constructor:
```julia
elseif matchmissing === :ignore
    if kind in [:left, :semi, :anti]
        dfr = dropmissing(dfr, right_on)
        dfr_on = dfr[!, right_on]
    elseif kind === :right
        dfl = dropmissing(dfl, left_on)
        dfl_on = dfl[!, left_on]
    elseif kind === :inner
        dfl = dropmissing(dfl, left_on)
        dfl_on = dfl[!, left_on]
        dfr = dropmissing(dfr, right_on)
        dfr_on = dfr[!, right_on]
    elseif kind === :outer 
        throw("matchmissing == :ignore with kind == :outer not allowed")
    else
        throw("matchmissing == :ignore not implemented for kind == $kind")
    end
```
I don't like solution with adding computation part to constructor but at least `matchmissing` logic is being kept in one place.

I was afraid that creating dataframes copies (by `dropmissing`) may blowup memory usage but I was unable to show it in benchmarks.  
I've tried setting `view=true` but it was slower and used more memory [probably due to this](https://docs.julialang.org/en/v1/manual/performance-tips/#Copying-data-is-not-always-bad).
Benchmark is attached in pdf.
[Ignorematch.pdf](https://github.com/JuliaData/DataFrames.jl/files/6290556/Ignorematch.pdf)
@bkamins  As you are the major contributor to joins, please tell my what do you think about this approach and I'll add tests, docs if you think it's acceptable.

**Comment by bkamins** [comment]
Thank you for working on this.

Throwing error should be:
```
throw(ArgumentError("matchmissing == :ignore with kind == :outer not allowed"))
```
but this is a minor thing.

I did not implement this PR immediately exactly because making it fast and not memory greedy requires benchmarking. My initial idea was:
1. use `completecases` instead of `dropmissing` and store the selector.
2. then materialize (as in your code) only `dfr_on` and `dfl_on` (as needed) with dropped cases
3. finally when composing the final table apply `completecases` to `dfr` and `dfl` along with the join condition

But this is complex.

Actually your proposal seems better, but I would do:
```
dfr = dropmissing(dfr, right_on, view=true)
dfr_on = select(dfr, right_on)
```
in this way `dfr` is a view (so it does not allocate much), but `dfr_on` is materialized (and most likely it is a narrow table). This should be both memory efficient (as materializing `dfr` can be prohibitive if this table is very wide) but at the same time fast to join (as only `dfr_on` is used in joining logic).

Could you please benchmark it and let me know.

One small comment. The code `dfr[!, right_on]` is correct but is unsafe. I preferred `select` over `getindex` as it clearly signals that we want to get a data frame (as `dfr[!, right_on]` could produce a vector; in this case it is ok as `right_on` is guaranteed to be a vector but in the future it might change so it is safer to use `select` which always produces a data frame).

**Comment by pstorozenko** [comment]
For some reason it is always more efficient to do
```
dfr = dropmissing(dfr, right_on)
dfr_on = select(dfr, right_on, copycols=false)
```
then
```
dfr = dropmissing(dfr, right_on, view=true)
dfr_on = select(dfr, right_on)
```
Benchmark here 
[Ignorematch_select.pdf](https://github.com/JuliaData/DataFrames.jl/files/6292288/Ignorematch_select.pdf)
I should add that there was always quite big variance in running times so sometimes `matchmissing = :error` appears slower then `:ignore`.
Do you have any specific benchmark scenarios in mind to check?

**Comment by bkamins** [comment]
This is super strange. I am getting different results of the core part that makes the difference between both options:
```
julia> df = DataFrame(rand(10^6, 5), :auto);

julia> idx = rand(1:10^6, 10^6);

julia> @btime dropmissing($df)[$idx, [1,2,3,4]];
  51.561 ms (99 allocations: 76.44 MiB)

julia> @btime dropmissing($df, view=true)[$idx, [1,2,3,4]];
  48.648 ms (75 allocations: 45.92 MiB)
```
and using a `view` is a bit faster (but this is not the part we are fighting for) and allocates much less (this is what we care about)

The difference is much bigger for wide tables:
```
```

**Comment by bkamins** [comment]
maybe open a PR with your changes and then I will profile both options? (or maybe you know how to do profiling to identify the source of the performance loss then please check and let me know)

Thank you very much for working on this.

**Comment by pstorozenko** [comment]
I wasn't interpolating dataframes in `@btime`, I reran experiments and results are the same, less allocs but more memory.

As this change is not very important/urgent right now, I'll try profiling myself to learn how to do it in julia and will let you know results later.

**Comment by bkamins** [comment]
Sure. The release timing is:
1. DataFrames.jl 1.0: April 2021
2. DataFrames.jl 1.1: late summer 2021

but I did not assume to have this feature in 1.0 release so there is no rush.

**Comment by bkamins** [comment]
I have renamed the option to `:notequal` as it is more explicit than `:ignore` (because we are not ignoring `missing` but treating them as not equal with this option).

**Comment by pstorozenko** [comment]
I created a PR with my changes as you asked: #2724 

After some profiling with `--track-allocation=user`:
```julia
df = DataFrame([rand(100000, 3) repeat(1:1000, 100)], :auto)
df_lm = DataFrame([rand(100000, 3) repeat([collect(1:500); repeat([missing], 500)], 100)], :auto)
innerjoin(df, df_lm, on=on, makeunique=true, matchmissing = :notequal_*)
```

Top 10 lines with memory allocation (memory is in MB) are:
```
10×4 DataFrame
 Row │ filename                           linenumber  mem_copy           mem_view          
     │ String                             Int64       Float64?           Float64?          
─────┼─────────────────────────────────────────────────────────────────────────────────────
   1 │ src/subdataframe/subdataframe.jl          170  missing                 76.2945
   2 │ src/join/core.jl                          466       49.5911            49.5911
   3 │ src/join/core.jl                          470       49.5911            49.5911
   4 │ src/abstractdataframe/abstractda…         847  missing                  1.14499
   5 │ src/join/composer.jl                       68        0.763336     missing           
   6 │ src/join/composer.jl                       70        0.381866     missing           
   7 │ src/join/core.jl                          381        0.381546           0.381546
   8 │ src/join/core.jl                          448        0.381546           0.381546
   9 │ src/join/core.jl                          454        0.381531           0.381531
  10 │ src/join/core.jl                          455        0.381531           0.381531
```
You'll probably have some intuition why that is the case.
What was suspicious for me is that I wasn't able to pinpoint lines responsible for allocation due to `dropmissing(df, on)`.
Please let me know what do you think about this.

**Comment by bkamins** [comment]
I will comment in the PR to keep things in one place.

**Comment by bkamins** [comment]
Fixed with https://github.com/JuliaData/DataFrames.jl/pull/2724. If there is anything left to add please comment.

**Comment by fkrauer** [comment]
Is there a way to use `outerjoin()` with `matchmissing = :notequal`? I have three dataframes, which I want to merge. They all have the same columns, some of which have missing values in all three dataframes. None of the rows in the three dataframes are identical, so the merge will in fact not merge any rows. I could probably also solve this by appending the three dataframes, but the order of the columns is not the same throughout the three dataframes. I think an outerjoin is the cleaner solution, but it cannot be done because  `matchmissing = :notequal` is not allowed at the moment. Is there a better way to solve this? Here is a MWE:

**Comment by fkrauer** [comment]
I figured it out:

**Comment by bkamins** [comment]
>  because `matchmissing = :notequal` is not allowed at the moment.

It is on purpose, as then you would have ambiguity in the output since `missing` is used as a fill value in column without match.

Now - you can also `append!` if you want it done in place:
```
julia> append!(df1, df2, cols=:union)
8×3 DataFrame
 Row │ y        x      z
     │ Int64?   Int64  Float64?
─────┼───────────────────────────
   1 │       1      1  missing
   2 │ missing      2  missing
   3 │       3      3  missing
   4 │       4      4  missing
   5 │       1      5  missing
   6 │       2      6  missing
   7 │ missing      7        5.0
   8 │       4      8       10.0
```

so I close it, but if you are convinced we need this option in `outerjoin` please comment and we can discuss.
